internal/cmd/perfcomp/main.go

```diff
@@ -0,0 +1,422 @@
+			"Error decoding raw data from version %q: %v",
+			version,
+			err,
+		)
+	}
+
+	return rawData, err
+}
+
+// Find the most recent stable region of the mainline version for a specific test/measurement
+func findLastStableRegion(project string, testname string, measurement string, coll *mongo.Collection) (*StableRegion, error) {
+	filter := bson.D{
+		{"time_series_info.project", project},
+		{"time_series_info.variant", "perf"},
+		{"time_series_info.task", "perf"},
+		{"time_series_info.test", testname},
+		{"time_series_info.measurement", measurement},
+		{"last", true},
+		{"contexts", bson.D{{"$in", bson.A{"GoDriver perf task"}}}},
+	}
+
+	findOptions := options.FindOne().SetSort(bson.D{{"end", -1}})
+
+	findCtx, cancel := context.WithTimeout(context.Background(), 180*time.Second)
+	defer cancel()
+
+	var sr *StableRegion
+	err := coll.FindOne(findCtx, filter, findOptions).Decode(&sr)
+	if err != nil {
+		return nil, err
+	}
+	return sr, nil
+}
+
+// For a specific test and measurement
+func getEnergyStatsForOneBenchmark(rd RawData, coll *mongo.Collection) ([]*EnergyStats, error) {
+	testname := rd.Info.TestName
+	var energyStats []*EnergyStats
+
+	for i := range rd.Rollups.Stats {
+		project := rd.Info.Project
+		measName := rd.Rollups.Stats[i].Name
+		measVal := rd.Rollups.Stats[i].Val
+
+		stableRegion, err := findLastStableRegion(project, testname, measName, coll)
+		if err != nil {
+			log.Fatalf(
+				"Error finding last stable region for test %q, measurement %q: %v",
+				testname,
+				measName,
+				err,
+			)
+		}
+
+		// The performance analyzer compares the measurement value from the patch to a stable region that succeeds the latest change point.
+		// For example, if there were 5 measurements since the last change point, then the stable region is the 5 latest values for the measurement.
+		stableRegionVec := mat.NewDense(len(stableRegion.Values), 1, stableRegion.Values)
+		measValVec := mat.NewDense(1, 1, []float64{measVal}) // singleton
+
+		estat, tstat, hscore, err := getEnergyStatistics(stableRegionVec, measValVec)
+		var zscore float64
+		var pChange float64
+		if err != nil {
+			log.Printf("Could not calculate energy stats for test %q, measurement %q: %v", testname, measName, err)
+			zscore = 0
+			pChange = 0
+		} else {
+			zscore = getZScore(measVal, stableRegion.Mean, stableRegion.Std)
+			pChange = getPercentageChange(measVal, stableRegion.Mean)
+		}
+
+		es := EnergyStats{
+			Project:         project,
+			Benchmark:       testname,
+			Measurement:     measName,
+			PatchVersion:    rd.Info.Version,
+			StableRegion:    *stableRegion,
+			MeasurementVal:  measVal,
+			PercentChange:   pChange,
+			EnergyStatistic: estat,
+			TestStatistic:   tstat,
+			HScore:          hscore,
+			ZScore:          zscore,
+		}
+		energyStats = append(energyStats, &es)
+	}
+
+	return energyStats, nil
+}
+
+func getEnergyStatsForAllBenchMarks(patchRawData []RawData, coll *mongo.Collection) ([]*EnergyStats, error) {
```

Comment:
Suggest adding a context to the function signature, as this is a blocking function.